src/ParserCombinators/Parser.hs

```diff
@@ -0,0 +1,122 @@
+{-# LANGUAGE GADTSyntax #-}
+{-# LANGUAGE InstanceSigs #-}
+
+module ParserCombinators.Parser
+  ( token
+  , satisfy
+  , eof
+  , char
+  , string
+  , between
+  , space
+  , spaces
+  , spaces1
+  , bool
+  , digit
+  , integer
+  , symbol
+  , many1
+  , try
+  , sepBy
+  , sepBy1
+  ) where
+
+import Control.Applicative (Alternative (..))
+import Control.Monad (void)
+import Data.Char (isAlphaNum, isDigit)
+import Data.Foldable (asum)
+
+import ParserCombinators.Datatypes
+
+-- | Parse a token/character using a custom error generator and condition
+token :: (i -> ParseErrorType i e) -> (i -> Bool) -> Parser i [] e i
+token mkErr predicate = Parser $ \input offset ->
+  case input of
+    [] -> Left $ pure $ ParseError offset EndOfInput
+    x : rest
+      | predicate x -> Right (succ offset, x, rest)
+      | otherwise -> Left $ pure $ ParseError offset $ mkErr x
```

Comment:
I recommend moving this and anything that actually touches (or needs to look at) the constructor of the `Parser` data type to the module where it's defined. You can then hide the constructor

The "types" module approach to structuring a program isn't good in general imo (e.g. for compilation speeds it doesn't scale well), and I also don't know what the benefits to doing it are.